core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -555,24 +479,43 @@ private object RTS {
                       io.value.ensuring(exitSupervision(io.supervisor))
 
                   case IO.Tags.Terminate =>
-                    val io = curIo.asInstanceOf[IO.Terminate]
+                    val io = curIo.asInstanceOf[IO.Terminate[E]]
 
-                    val finalizer = interruptStack
+                    val finalizer = interruptStack(io.cause.checkedFirst.isDefined)
 
-                    if (finalizer eq null) {
-                      // No finalizers, simply produce error:
-                      val causes    = status.get.terminationCauses.getOrElse(Nil)
-                      val defects   = status.get.defects
-                      val allCauses = io.causes ++ causes ++ defects
+                    if (stack.isEmpty) {
+                      // Error not caught, stack is empty:
+                      if (finalizer eq null) {
+                        // No finalizer, so immediately produce the error.
+                        curIo = null
+
+                        val cause = (io.cause, status.get.terminationCause) match {
```

Comment:
First, `terminationCause` should be replaced by `interrupted: Boolean` inside the fiber state. This is because `interrupt` will _no longer_ accept a list of throwables. That is, interruption will always carry no additional information on the cause of interruption.

This means you can replace all of this `cause` code by something like:

```scala
val cause = if (status.get.interrupted) io.cause ++ Cause.interrupt else io.cause
```

I see also the old code used `status.get.defects` to accumulate errors during finalization. I think you are now accumulating them into `terminationCause`. We need to change that, ideally by no longer storing finalization errors into the ref, but just accumulating them directly.

However, in this branch, there are no finalizers, so the above simplification should be enough.

Reply:
⚡️